core/src/test/java/org/apache/iceberg/TestPartitionSpecUpdate.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestPartitionSpecUpdate extends TableTestBase {
+
+  private int[] expectedFieldIds;
+
+  private String[] expectedSpecs;
+
+  @Parameterized.Parameters
+  public static Object[][] parameters() {
+    return new Object[][] {
+        new Object[] { 1, new int[]{ 1000, 1001, 1001, 1000, 1000 },
+            new String[]{"  1000: data_bucket_removed: void(2)\n", "  1002: id_bucket_removed: void(1)\n"} },
+        new Object[] { 2, new int[]{ 1001, 1000, 1001, 1002, 1002 }, new String[]{"", ""} },
+    };
+  }
+
+  public TestPartitionSpecUpdate(int formatVersion, int[] expectedFieldIds, String[] expectedSpecs) {
+    super(formatVersion);
+    this.expectedFieldIds = expectedFieldIds;
+    this.expectedSpecs = expectedSpecs;
+  }
+
+  @Test
+  public void testCommitUpdatedSpec() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addBucketField("id", 8)
+        .addBucketField("data", 16)
+        .commit();
+
+    Assert.assertEquals("[\n  " +
+        expectedFieldIds[0] + ": id_bucket: bucket[8](1)\n  " +
+        expectedFieldIds[1] + ": data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(expectedFieldIds[2], table.spec().lastAssignedFieldId());
```

Comment:
If I'm reading this correctly, then either `id_bucket` or `data_bucket` is assigned ID 1000, depending on the table's format version. But 1000 was the ID assigned to the original data_bucket field and it should not be reused. That would break the `files` table by mixing `id_bucket` and the original `data_bucket` values together.

Reply:
I think the problem is that `clear` is not replacing the existing fields with `alwaysNull` as a placeholder for the existing fields.